python-package/setup.py

```diff
@@ -118,8 +118,22 @@ def compile_cpp(use_mingw=False, use_gpu=False, use_mpi=False, nomp=False,
         cmake_cmd.append("-DUSE_MPI=ON")
     if nomp:
         cmake_cmd.append("-DUSE_OPENMP=OFF")
+    if system() == 'Darwin' and not nomp:
+        cc = os.environ.get('CC')
+        cxx = os.environ.get('CXX')
+        if not (cc and cc.startswith('gcc') and cxx and cxx.startswith('g++')):
+            # Apple Clang
+            # https://github.com/Homebrew/homebrew-core/pull/20589
+            cmake_cmd.extend([
+                '-DOpenMP_C_FLAGS="-Xpreprocessor -fopenmp -I/usr/local/opt/libomp/include"',
+                '-DOpenMP_C_LIB_NAMES="omp"',
+                '-DOpenMP_CXX_FLAGS="-Xpreprocessor -fopenmp -I/usr/local/opt/libomp/include"',
+                '-DOpenMP_CXX_LIB_NAMES="omp"',
+                '-DOpenMP_"omp"_LIBRARY=/usr/local/opt/libomp/lib/libomp.dylib'
```

Comment:
I don't really know why this is `-DOpenMP_"omp"_LIBRARY`, not `-DOpenMP_omp_LIBRARY` 🤔
However, `python setup.py install` works only when ``-DOpenMP_"omp"_LIBRARY``.
If `-DOpenMP_omp_LIBRARY`, the bellow errors will occur:

```
-- The C compiler identification is AppleClang 10.0.0.10001145
-- The CXX compiler identification is AppleClang 10.0.0.10001145
-- Check for working C compiler: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/cc
-- Check for working C compiler: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/cc -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Detecting C compile features
-- Detecting C compile features - done
-- Check for working CXX compiler: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/c++
-- Check for working CXX compiler: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/c++ -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
CMake Error at /usr/local/Cellar/cmake/3.13.2/share/cmake/Modules/FindPackageHandleStandardArgs.cmake:137 (message):
  Could NOT find OpenMP_C (missing: OpenMP_"omp"_LIBRARY)
Call Stack (most recent call first):
  /usr/local/Cellar/cmake/3.13.2/share/cmake/Modules/FindPackageHandleStandardArgs.cmake:378 (_FPHSA_FAILURE_MESSAGE)
  /usr/local/Cellar/cmake/3.13.2/share/cmake/Modules/FindOpenMP.cmake:473 (find_package_handle_standard_args)
  CMakeLists.txt:78 (find_package)


-- Configuring incomplete, errors occurred!
See also "/Users/henry/.ghq/github.com/henry0312/LightGBM/python-package/build_cpp/CMakeFiles/CMakeOutput.log".
See also "/Users/henry/.ghq/github.com/henry0312/LightGBM/python-package/build_cpp/CMakeFiles/CMakeError.log".
```

Would anyone tell me what happens in `setup.py`, please.

Reply:
Please try to remove **all** quotes from the command.

Reply:
I'm sorry, I don't understand what you mean 😢
Which quotes should I remove from the command?

Reply:
I suppose **all**. For example:
```
'-DOpenMP_CXX_FLAGS=-Xpreprocessor -fopenmp -I/usr/local/opt/libomp/include',
'-DOpenMP_CXX_LIB_NAMES=omp',
```

Reply:
It's only my assumption, based on the fact that we do not use `shell=True`. Sorry, that I don't check my hypotheses by myself, but I don't have access to Mojave.

Reply:
```python
            cmake_cmd.extend([
                '-DOpenMP_C_FLAGS=-Xpreprocessor -fopenmp -I/usr/local/opt/libomp/include',
                '-DOpenMP_C_LIB_NAMES=omp',
                '-DOpenMP_CXX_FLAGS=-Xpreprocessor -fopenmp -I/usr/local/opt/libomp/include',
                '-DOpenMP_CXX_LIB_NAMES=omp',
                '-DOpenMP_omp_LIBRARY=/usr/local/opt/libomp/lib/libomp.dylib'
            ])
```

works!
Would you gime me some documetns that tell me why?

Reply:
Sorry, it was my experience from VS command building (you can see it above in `setup.py`) more than year ago. I may be wrong, but my intuition is in that quotes are needed only for interaction through a shell to escape whitespaces. However, we are using `subprocess` without `shell=True` (which is insecure, BTW, as Python docs says), so all quotes are goes directly as part of an argument.

I'll try to find some official information about it for you

Reply:
@henry0312 I've tried to google, but unfortunately without any luck. Sorry, it was more that year ago and I can't recollect in memory what I found that made me omit quotes. 😢 
Here's only partially related answer on stackoverflow: https://stackoverflow.com/a/14632378